daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,272 @@
+          case Norm.Exe(exe, subs) =>
+            pushNorms(s, subs) { (s, children) =>
+              val node = exe.copy(children = ImmArray(children))
+              s.push(me, node)(k)
+            }
+        }
+      }
+    }
+  }
+
+  private def pushRoll[R](s: State, x: Norm.Roll)(k: (State, Nid) => Tramp[R]): Tramp[R] = {
+    s.next { (s, me) =>
+      x match {
+        case Norm.Roll1(act) =>
+          pushAct(s, act) { (s, child) =>
+            val node = NodeRollback(children = ImmArray(List(child)))
+            s.push(me, node)(k)
+          }
+        case Norm.Roll2(h, m, t) =>
+          pushAct(s, h) { (s, hh) =>
+            pushNorms(s, m) { (s, mm) =>
+              pushAct(s, t) { (s, tt) =>
+                val children = List(hh) ++ mm ++ List(tt)
+                val node = NodeRollback(children = ImmArray(children))
+                s.push(me, node)(k)
+              }
+            }
+          }
+      }
+    }
+  }
+
+  private def pushNorm[R](s: State, x: Norm)(k: (State, Nid) => Tramp[R]): Tramp[R] = {
+    x match {
+      case act: Norm.Act => pushAct(s, act)(k)
+      case roll: Norm.Roll => pushRoll(s, roll)(k)
+    }
+  }
+
+  private def pushNorms[R](s: State, xs: List[Norm])(
+      k: (State, List[Nid]) => Tramp[R]
+  ): Tramp[R] = {
+    Bounce { () =>
+      xs match {
+        case Nil => k(s, Nil)
+        case x :: xs =>
+          pushNorm(s, x) { (s, y) =>
+            pushNorms(s, xs) { (s, ys) =>
+              Bounce { () =>
+                k(s, y :: ys)
+              }
+            }
+          }
+      }
+    }
+  }
+
+  // Types which ensure we can only represent the properly normalized cases.
+  private object Canonical {
+
+    // A properly normalized Tx/node
+    sealed trait Norm
+    object Norm {
+
+      // A non-rollback tx/node
+      sealed trait Act extends Norm
+      final case class Leaf(node: LeafNode) extends Act
+      final case class Exe(node: ExeNode, children: List[Norm]) extends Act
+
+      // A *normalized* rollback tx/node. 2 cases:
+      // - rollback containing a single non-rollback tx/node.
+      // - rollback of 2 or more tx/nodes, such that first and last are not rollbacks.
+      sealed trait Roll extends Norm
+      final case class Roll1(act: Act) extends Roll
+      final case class Roll2(head: Act, middle: List[Norm], tail: Act) extends Roll
+    }
+
+    // Case analysis on a list of Norms, distinuishing: Empty, Single and Multi forms
+    // The Multi form separes the head and tail element for the middle-list.
+    sealed trait Case
+    object Case {
+      final case object Empty extends Case
+      final case class Single(n: Norm) extends Case
+      final case class Multi(h: Norm, m: List[Norm], t: Norm) extends Case
+    }
+
+    def caseNorms(xs: List[Norm]): Case = {
+      xs.length match {
+        case 0 => Case.Empty
+        case 1 => Case.Single(xs(0))
+        case n => Case.Multi(xs(0), xs.slice(1, n - 1), xs(n - 1))
```

Comment:
```suggestion
      xs match {
        case Nil => Case.Empty
        case x0 :: Nil => Case.Single(h)
        case x0 :: x1 :: tail  => Case.Multi(x0, List(x1), tail)
```

Reply:
That’s not the same. Before we had first, middle, last. Now you have first, second, rest.

Reply:
indeed.
We should then use `ImmArray` or `Vector`.